### PR TITLE
BAVL-235 initial commit adding localstack for integration testing of events/messages.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
Adds local stack test containers.

Note the tests using this only run locally and are disabled on the build pipeline due to very occasional timing issues.  We do not want to stop the build going out because of this.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -17,8 +17,13 @@ parameters:
 jobs:
   validate:
     executor:
-      name: hmpps/java
-      tag: << pipeline.parameters.java-version >>
+      name: hmpps/java_localstack_postgres
```
Comment (matt-crowson-moj) [inline .circleci/config.yml:20]: We are not using a postgres container but this is the only executor present.

It does not interfere with the build but perhaps we should move to using a postgres container as well.  Would do this in a separate PR though.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/config/FeatureSwitchesTest.kt [hunk 2]
```diff
@@ -6,14 +6,24 @@ import org.junit.jupiter.api.Nested
...
 import org.springframework.test.context.TestPropertySource
 import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.helper.isBool
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.service.events.InboundEventsListener
+import uk.gov.justice.hmpps.sqs.HmppsQueueService
 
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
 @ActiveProfiles("test")
 class FeatureSwitchesTest {
 
+  // Beans are mocked out so as not to interfere with the running of the tests. We do not want/need them on the context.
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/config/FeatureSwitchesTest.kt:20]: This is a side effect of the InboundEventsListener no longer being filtered out in tests.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/events/InboundEventsListener.kt [hunk 3]
```diff
@@ -14,7 +14,7 @@ import org.springframework.stereotype.Component
 import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.config.Feature
 import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.config.FeatureSwitches
 
-@Profile("!test && !local")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/events/InboundEventsListener.kt:17]: We now want this to be added to the context at startup for integration tests.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/resource/VideoLinkBookingIntegrationTest.kt [hunk 4]
```diff
@@ -64,20 +66,29 @@ import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.service.CourtBookingR
...
 class VideoLinkBookingIntegrationTest : IntegrationTestBase() {
 
-  @MockBean
+  @SpyBean
   private lateinit var outboundEventsPublisher: OutboundEventsPublisher
 
   private val domainEventCaptor = argumentCaptor<DomainEvent<*>>()
 
+  @MockBean
+  private lateinit var manageExternalAppointmentsService: ManageExternalAppointmentsService
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/resource/VideoLinkBookingIntegrationTest.kt:90]: This is being mocked because now we are getting events processed it is also trying to call the external services.  It is not being tested at the moment though here.

We could assert the various methods, create, cancel etc are called.

File: .circleci/config.yml (hunk 1 above)
Comment (timharrison-moj) [reply .circleci/config.yml:20]: If postgresql for integration tests, best done as a TestContainers? So started automatically for int tests.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/resource/VideoLinkBookingIntegrationTest.kt (hunk 4 above)
Comment (timharrison-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/resource/VideoLinkBookingIntegrationTest.kt:90]: Alternative is a mock service for each external service, but probably enough to know that the external service methods are being triggered by the inbound events.

**Review by timharrison-moj — DISMISSED** [review]
LGTM. 
I'll need to study a bit more to work out any further impact on int tests.

**Comment by matt-crowson-moj** [comment]
> LGTM. I'll need to study a bit more to work out any further impact on int tests.

Ideally I like to only spin up localstack for those tests that need/use it.  This requires some refactoring though.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/resource/VideoLinkBookingIntegrationTest.kt (hunk 4 above)
Comment (matt-crowson-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/resource/VideoLinkBookingIntegrationTest.kt:90]: Yeah I know.  It's a bit of a balancing act as in how far do you take it?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/TestContainersIntegrationTestBase.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.service.events.Message
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.service.events.MessageAttributes
+import uk.gov.justice.hmpps.sqs.HmppsQueue
+import uk.gov.justice.hmpps.sqs.HmppsQueueService
+import uk.gov.justice.hmpps.sqs.countAllMessagesOnQueue
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+@ActiveProfiles("test-localstack", inheritProfiles = false)
+abstract class TestContainersIntegrationTestBase : IntegrationTestBase() {
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/TestContainersIntegrationTestBase.kt:31]: This might be better renamed to something else?  SQS/SNS/Message based.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/resource/InboundEventsIntegrationTest.kt [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.integration.wiremock.TEST_USERNAME
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.integration.wiremock.TEST_USER_EMAIL
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.request.CreateVideoBookingRequest
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.repository.VideoBookingRepository
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.service.events.ManageExternalAppointmentsService
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.service.events.PrisonerReleasedEvent
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.service.events.ReleaseInformation
+import java.time.LocalTime
+
+@DisabledIfEnvironmentVariable(named = "CIRCLECI", matches = "true")
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/resource/InboundEventsIntegrationTest.kt:29]: This env var already exists in Circle so is available when the pipeline build/check runs.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/events/handlers/PrisonerReleasedEventHandler.kt [hunk 7]
```diff
@@ -29,7 +29,12 @@ class PrisonerReleasedEventHandler(
   }
 
   private fun cancelFutureBookingsFor(prisonerNumber: String) {
+//    prisonAppointmentRepository.findByPrisonerNumber(prisonerNumber)
     prisonAppointmentRepository.findPrisonerPrisonAppointmentsAfter(prisonerNumber, LocalDate.now(), LocalTime.now())
+      .ifEmpty {
+        log.info("No future bookings currently exists for prisoner $prisonerNumber")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/events/handlers/PrisonerReleasedEventHandler.kt:35]: Could probably remove this?

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/events/handlers/PrisonerReleasedEventHandler.kt:35]: Might come in handy? Maybe leave for now, and if ceases to be useful remove later?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/TestContainersIntegrationTestBase.kt (hunk 5 above)
Comment (timharrison-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/TestContainersIntegrationTestBase.kt:31]: Yeah, TestContainersIntegrationTestBase does not really give a clue that its for localstack-based tests.

**Review by timharrison-moj — COMMENTED** [review]
LGTM. Just the base class to rename and should be good to go.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/TestContainersIntegrationTestBase.kt (hunk 5 above)
Comment (matt-crowson-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/integration/TestContainersIntegrationTestBase.kt]: Done

**Review by timharrison-moj — APPROVED** [review]
LGTM
